src/java.base/share/classes/java/lang/foreign/Arena.java

```diff
@@ -0,0 +1,167 @@
+     * {@snippet lang = java:
+     * MemorySegment.allocateNative(bytesSize, byteAlignment, scope());
+     *}
+     * More generally implementations of this method must return a native segment featuring the requested size,
+     * and that is compatible with the provided alignment constraint. Furthermore, for any two segments
+     * {@code S1, S2} returned by this method, the following invariant must hold:
+     *
+     * {@snippet lang = java:
+     * S1.overlappingSlice(S2).isEmpty() == true
+     *}
+     *
+     * @param byteSize the size (in bytes) of the off-heap memory block backing the native memory segment.
+     * @param byteAlignment the alignment constraint (in bytes) of the off-heap region of memory backing the native memory segment.
+     * @return a new native memory segment.
+     * @throws IllegalArgumentException if {@code bytesSize < 0}, {@code alignmentBytes <= 0}, or if {@code alignmentBytes}
+     * is not a power of 2.
+     * @throws IllegalStateException if the arena has already been {@linkplain #close() closed}.
+     * @throws WrongThreadException if this method is called from a thread {@code T},
+     * such that {@code scope().isAccessibleBy(T) == false}.
+     * @see MemorySegment#allocateNative(long, long, SegmentScope)
+     */
+    @Override
+    default MemorySegment allocate(long byteSize, long byteAlignment) {
+        return MemorySegment.allocateNative(byteSize, byteAlignment, scope());
+    }
+
+    /**
+     * {@return the arena scope}
+     */
+    SegmentScope scope();
+
+    /**
+     * Closes this arena. If this method completes normally, the arena scope is no longer {@linkplain SegmentScope#isAlive() alive},
+     * and all the memory segments associated with it can no longer be accessed. Furthermore, any off-heap region of memory backing the
+     * segments associated with that scope are also released.
+     *
+     * @apiNote This operation is not idempotent; that is, closing an already closed arena <em>always</em> results in an
+     * exception being thrown. This reflects a deliberate design choice: arena state transitions should be
+     * manifest in the client code; a failure in any of these transitions reveals a bug in the underlying application
+     * logic.
+     *
+     * @see SegmentScope#isAlive()
+     *
+     * @throws IllegalStateException if the arena has already been closed.
+     * @throws IllegalStateException if the arena scope is {@linkplain SegmentScope#whileAlive(Runnable) kept alive}.
+     * @throws WrongThreadException if this method is called from a thread {@code T},
+     * such that {@code isCloseableBy(T) == false}.
+     */
+    @Override
+    void close();
+
+    /**
+     * {@return {@code true} if the provided thread can close this arena}
+     * @param thread the thread to be tested.
+     */
+    boolean isCloseableBy(Thread thread);
+
+    /**
+     * {@return a new confined arena}
```

Comment:
For completeness, this should probably say "a new confined arena owned by the current thread".